### ISSUE TITLE
Performance degradation with custom comparator

### ISSUE THREAD

**Issue by jvilk** [andrasq/node-qheap#1]
I figured you might want to know about this. I modified the benchmark script to use objects and a custom comparator, as that is how I planned to use a priority queue. `qheap` ended up being surprisingly slow!

Modified test:

``` js
/* performance benchmark */
/* This script expects node.js */

'use strict';

var FastPriorityQueue = require('../FastPriorityQueue.js');
var PriorityQueue = require('js-priority-queue');
var Heap = require('heap');
var Benchmark = require('benchmark');
var os = require('os');
var binaryheapx = require('binaryheapx').Constructor;
var pq = require('priority_queue');
var jsHeap = require('js-heap');
var qp = require('queue-priority');
var priorityqueuejs = require('priorityqueuejs');
var qheap = require('qheap');
var BinaryHeap = require('yabh');

function BoxedNumber(i) {
  this.i = i;
}

// very fast semi-random function
function rand(i) {
  i = i + 10000;
  i = i ^ (i << 16);
  i = (i >> 5) ^ i ;
  return new BoxedNumber(i & 0xFF);
}
function defaultcomparator(a, b) {
  return a.i < b.i;
}
function subtractComparator(a, b) {
  return a.i - b.i;
}
function weightFcn(x) { return x.i; }

function basictest() {
  var b1 = new FastPriorityQueue(defaultcomparator);
  var b2 = new Heap(subtractComparator);
  for (var i = 0 ; i < 2000  ; i++) {
    b1.add(rand(i));
    b2.push(rand(i));
  }
  for (i = 128 ; i < 128 * 10  ; i++) {
    b1.add(rand(i));
    b2.push(rand(i));
    var x1 = b1.poll();
    var x2 = b2.pop();
    if (x1.i != x2.i) throw 'bug ' + x1 + ' ' + x2;
  }

}

function QueueEnqueueBench() {
  console.log('starting dynamic queue/enqueue benchmark');
  var suite = new Benchmark.Suite();
  // add tests
  var b = new BinaryHeap(weightFcn, subtractComparator);
  for (var i = 0 ; i < 128  ; i++) {
    b.push(rand(i));
  }
  for (i = 128 ; i < 128 * 10  ; i++) {
    b.push(rand(i));
    b.pop();
  }

  var ms = suite
    .add('FastPriorityQueue', function() {
      var b = new FastPriorityQueue(defaultcomparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.add(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.add(rand(i));
        b.poll();
      }
      return b;
    })
   .add('js-priority-queue', function() {
      var b = new PriorityQueue({comparator: subtractComparator});
      for (var i = 0 ; i < 128  ; i++) {
        b.queue(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.queue(rand(i));
        b.dequeue();
      }
      return b;
    })
    .add('heap.js', function() {
      var b = new Heap(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('binaryheapx', function() {
      var b = new binaryheapx(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.add(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.add(rand(i));
        b.pop();
      }
      return b;
    })
    .add('priority_queue', function() {
      var b = new pq.PriorityQueue(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.shift();
      }
      return b;
    })
    .add('js-heap', function() {
      var b = new jsHeap(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('queue-priority', function() {
      var b = new qp(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('priorityqueuejs', function() {
      var b = new priorityqueuejs(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.enq(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.enq(rand(i));
        b.deq();
      }
      return b;
    })
    .add('qheap', function() {
      var b = new qheap({ compar: subtractComparator });
      for (var i = 0 ; i < 128  ; i++) {
        b.insert(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.insert(rand(i));
        b.remove();
      }
      return b;
    })
    .add('yabh', function() {
      var b = new BinaryHeap(weightFcn, subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    // add listeners
    .on('cycle', function(event) {
      console.log(String(event.target));
    })
    .on('complete', function() {
      console.log('Fastest is ' + this.filter('fastest').pluck('name'));
    })
    // run async
    .run({'async': false});
}

var main = function() {
  basictest();
  console.log('Platform: ' + process.platform + ' ' + os.release() + ' ' + process.arch);
  console.log(os.cpus()[0]['model']);
  console.log('Node version ' + process.versions.node + ', v8 version ' + process.versions.v8);
  console.log();
  console.log('Comparing against: ');
  console.log('js-priority-queue: https://github.com/adamhooper/js-priority-queue');
  console.log('heap.js: https://github.com/qiao/heap.js');
  console.log('binaryheapx: https://github.com/xudafeng/BinaryHeap');
  console.log('priority_queue: https://github.com/agnat/js_priority_queue');
  console.log('js-heap: https://github.com/thauburger/js-heap');
  console.log('queue-priority: https://github.com/augustohp/Priority-Queue-NodeJS');
  console.log('priorityqueuejs: https://github.com/janogonzalez/priorityqueuejs');
  console.log('qheap: https://github.com/andrasq/node-qheap');
  console.log('yabh: https://github.com/jmdobry/yabh');
  console.log('');
  QueueEnqueueBench();
  console.log('');
};

if (require.main === module) {
  main();
}
```

Results:

```
Platform: darwin 15.5.0 x64
Intel(R) Core(TM) i5-5287U CPU @ 2.90GHz
Node version 6.2.2, v8 version 5.0.71.52

Comparing against:
js-priority-queue: https://github.com/adamhooper/js-priority-queue
heap.js: https://github.com/qiao/heap.js
binaryheapx: https://github.com/xudafeng/BinaryHeap
priority_queue: https://github.com/agnat/js_priority_queue
js-heap: https://github.com/thauburger/js-heap
queue-priority: https://github.com/augustohp/Priority-Queue-NodeJS
priorityqueuejs: https://github.com/janogonzalez/priorityqueuejs
qheap: https://github.com/andrasq/node-qheap
yabh: https://github.com/jmdobry/yabh

starting dynamic queue/enqueue benchmark
FastPriorityQueue x 12,094 ops/sec ±0.46% (97 runs sampled)
js-priority-queue x 6,884 ops/sec ±0.46% (97 runs sampled)
heap.js x 10,059 ops/sec ±1.02% (87 runs sampled)
binaryheapx x 6,130 ops/sec ±0.44% (97 runs sampled)
priority_queue x 2,078 ops/sec ±2.20% (92 runs sampled)
js-heap x 208 ops/sec ±1.06% (84 runs sampled)
queue-priority x 471 ops/sec ±0.52% (94 runs sampled)
priorityqueuejs x 4,916 ops/sec ±0.48% (98 runs sampled)
qheap x 2,083 ops/sec ±0.47% (97 runs sampled)
yabh x 7,693 ops/sec ±0.45% (97 runs sampled)
Fastest is FastPriorityQueue
```

Oddly enough, I get worse results when I use **numbers & a custom comparator**:

``` js
/* performance benchmark */
/* This script expects node.js */

'use strict';

var FastPriorityQueue = require('../FastPriorityQueue.js');
var PriorityQueue = require('js-priority-queue');
var Heap = require('heap');
var Benchmark = require('benchmark');
var os = require('os');
var binaryheapx = require('binaryheapx').Constructor;
var pq = require('priority_queue');
var jsHeap = require('js-heap');
var qp = require('queue-priority');
var priorityqueuejs = require('priorityqueuejs');
var qheap = require('qheap');
var BinaryHeap = require('yabh');

// very fast semi-random function
function rand(i) {
  i = i + 10000;
  i = i ^ (i << 16);
  i = (i >> 5) ^ i ;
  return i & 0xFF;
}
function defaultcomparator(a, b) {
  return a < b;
}
function subtractComparator(a, b) {
  return a - b;
}
function weightFcn(x) { return x; }

function basictest() {
  var b1 = new FastPriorityQueue(defaultcomparator);
  var b2 = new Heap(subtractComparator);
  for (var i = 0 ; i < 2000  ; i++) {
    b1.add(rand(i));
    b2.push(rand(i));
  }
  for (i = 128 ; i < 128 * 10  ; i++) {
    b1.add(rand(i));
    b2.push(rand(i));
    var x1 = b1.poll();
    var x2 = b2.pop();
    if (x1 != x2) throw 'bug ' + x1 + ' ' + x2;
  }

}

function QueueEnqueueBench() {
  console.log('starting dynamic queue/enqueue benchmark');
  var suite = new Benchmark.Suite();
  // add tests
  var b = new BinaryHeap(weightFcn, subtractComparator);
  for (var i = 0 ; i < 128  ; i++) {
    b.push(rand(i));
  }
  for (i = 128 ; i < 128 * 10  ; i++) {
    b.push(rand(i));
    b.pop();
  }

  var ms = suite
    .add('FastPriorityQueue', function() {
      var b = new FastPriorityQueue(defaultcomparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.add(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.add(rand(i));
        b.poll();
      }
      return b;
    })
   .add('js-priority-queue', function() {
      var b = new PriorityQueue({comparator: subtractComparator});
      for (var i = 0 ; i < 128  ; i++) {
        b.queue(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.queue(rand(i));
        b.dequeue();
      }
      return b;
    })
    .add('heap.js', function() {
      var b = new Heap(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('binaryheapx', function() {
      var b = new binaryheapx(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.add(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.add(rand(i));
        b.pop();
      }
      return b;
    })
    .add('priority_queue', function() {
      var b = new pq.PriorityQueue(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.shift();
      }
      return b;
    })
    .add('js-heap', function() {
      var b = new jsHeap(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('queue-priority', function() {
      var b = new qp(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    .add('priorityqueuejs', function() {
      var b = new priorityqueuejs(subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.enq(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.enq(rand(i));
        b.deq();
      }
      return b;
    })
    .add('qheap', function() {
      var b = new qheap({ compar: subtractComparator });
      for (var i = 0 ; i < 128  ; i++) {
        b.insert(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.insert(rand(i));
        b.remove();
      }
      return b;
    })
    .add('yabh', function() {
      var b = new BinaryHeap(weightFcn, subtractComparator);
      for (var i = 0 ; i < 128  ; i++) {
        b.push(rand(i));
      }
      for (i = 128 ; i < 128 * 10  ; i++) {
        b.push(rand(i));
        b.pop();
      }
      return b;
    })
    // add listeners
    .on('cycle', function(event) {
      console.log(String(event.target));
    })
    .on('complete', function() {
      console.log('Fastest is ' + this.filter('fastest').pluck('name'));
    })
    // run async
    .run({'async': false});
}

var main = function() {
  basictest();
  console.log('Platform: ' + process.platform + ' ' + os.release() + ' ' + process.arch);
  console.log(os.cpus()[0]['model']);
  console.log('Node version ' + process.versions.node + ', v8 version ' + process.versions.v8);
  console.log();
  console.log('Comparing against: ');
  console.log('js-priority-queue: https://github.com/adamhooper/js-priority-queue');
  console.log('heap.js: https://github.com/qiao/heap.js');
  console.log('binaryheapx: https://github.com/xudafeng/BinaryHeap');
  console.log('priority_queue: https://github.com/agnat/js_priority_queue');
  console.log('js-heap: https://github.com/thauburger/js-heap');
  console.log('queue-priority: https://github.com/augustohp/Priority-Queue-NodeJS');
  console.log('priorityqueuejs: https://github.com/janogonzalez/priorityqueuejs');
  console.log('qheap: https://github.com/andrasq/node-qheap');
  console.log('yabh: https://github.com/jmdobry/yabh');
  console.log('');
  QueueEnqueueBench();
  console.log('');
};

if (require.main === module) {
  main();
}
```

Results:

```
Platform: darwin 15.5.0 x64
Intel(R) Core(TM) i5-5287U CPU @ 2.90GHz
Node version 6.2.2, v8 version 5.0.71.52

Comparing against:
js-priority-queue: https://github.com/adamhooper/js-priority-queue
heap.js: https://github.com/qiao/heap.js
binaryheapx: https://github.com/xudafeng/BinaryHeap
priority_queue: https://github.com/agnat/js_priority_queue
js-heap: https://github.com/thauburger/js-heap
queue-priority: https://github.com/augustohp/Priority-Queue-NodeJS
priorityqueuejs: https://github.com/janogonzalez/priorityqueuejs
qheap: https://github.com/andrasq/node-qheap
yabh: https://github.com/jmdobry/yabh

starting dynamic queue/enqueue benchmark
FastPriorityQueue x 24,827 ops/sec ±1.54% (96 runs sampled)
js-priority-queue x 13,009 ops/sec ±1.25% (96 runs sampled)
heap.js x 17,146 ops/sec ±0.33% (98 runs sampled)
binaryheapx x 9,459 ops/sec ±0.36% (94 runs sampled)
priority_queue x 2,467 ops/sec ±1.07% (95 runs sampled)
js-heap x 186 ops/sec ±2.15% (81 runs sampled)
queue-priority x 4.98 ops/sec ±1.99% (17 runs sampled)
priorityqueuejs x 7,913 ops/sec ±0.28% (100 runs sampled)
qheap x 984 ops/sec ±6.92% (90 runs sampled)
yabh x 13,355 ops/sec ±0.89% (92 runs sampled)
Fastest is FastPriorityQueue
```

**Comment by andrasq** [comment]
indeed; thanks for pointing it out!  I was able to reproduce the slowdown.
It seems node does not optimize a general-purpose comparator as well as the default
built-in "comes before" test.

I exposed the internal test function as a new constructor option `comparBefore`; using
comparBefore instead of compar should run much faster.  Bumped to version 1.3.0.

**Comment by andrasq** [comment]
Fixed in 1.3.0
